src/Cli/Microsoft.DotNet.FileBasedPrograms/FileLevelDirectiveHelpers.cs

```diff
@@ -0,0 +1,621 @@
+            {
+                if (!Fill(ref result.Leading, triviaList, i)) break;
+            }
+
+            for (int i = index + 1; i < triviaList.Count; i++)
+            {
+                if (!Fill(ref result.Trailing, triviaList, i)) break;
+            }
+
+            return result;
+
+            static bool Fill(ref WhiteSpaceInfo info, in SyntaxTriviaList triviaList, int index)
+            {
+                var trivia = triviaList[index];
+                if (trivia.IsKind(SyntaxKind.EndOfLineTrivia))
+                {
+                    info.LineBreaks += 1;
+                    info.TotalLength += trivia.FullSpan.Length;
+                    return true;
+                }
+
+                if (trivia.IsKind(SyntaxKind.WhitespaceTrivia))
+                {
+                    info.TotalLength += trivia.FullSpan.Length;
+                    return true;
+                }
+
+                return false;
+            }
+        }
+    }
+}
+
+internal readonly record struct SourceFile(string Path, SourceText Text)
+{
+    public static SourceFile Load(string filePath)
+    {
+        using var stream = File.OpenRead(filePath);
+        return new SourceFile(filePath, SourceText.From(stream, Encoding.UTF8));
+    }
+
+    public SourceFile WithText(SourceText newText)
+    {
+        return new SourceFile(Path, newText);
+    }
+
+    public void Save()
+    {
+        using var stream = File.Open(Path, FileMode.Create, FileAccess.Write);
+        using var writer = new StreamWriter(stream, Encoding.UTF8);
+        Text.Write(writer);
+    }
+
+    public FileLinePositionSpan GetFileLinePositionSpan(TextSpan span)
+    {
+        return new FileLinePositionSpan(Path, Text.Lines.GetLinePositionSpan(span));
+    }
+
+    public string GetLocationString(TextSpan span)
+    {
+        var positionSpan = GetFileLinePositionSpan(span);
+        return $"{positionSpan.Path}({positionSpan.StartLinePosition.Line + 1})";
+    }
+}
+
+internal static partial class Patterns
+{
+    public static Regex Whitespace { get; } = new Regex("""\s+""", RegexOptions.Compiled);
+
+    public static Regex DisallowedNameCharacters { get; } = new Regex("""[\s@=/]""", RegexOptions.Compiled);
+
+    public static Regex EscapedCompilerOption { get; } = new Regex("""^/\w+:".*"$""", RegexOptions.Compiled | RegexOptions.Singleline);
+}
+
+internal struct WhiteSpaceInfo
+{
+    public int LineBreaks;
+    public int TotalLength;
+}
+
+/// <summary>
+/// Represents a C# directive starting with <c>#:</c> (a.k.a., "file-level directive").
+/// Those are ignored by the language but recognized by us.
+/// </summary>
+internal abstract class CSharpDirective(in CSharpDirective.ParseInfo info)
+{
+    public ParseInfo Info { get; } = info;
+
+    public readonly struct ParseInfo
+    {
+        /// <summary>
+        /// Span of the full line including the trailing line break.
+        /// </summary>
+        public required TextSpan Span { get; init; }
```

Comment:
I don't understand moving off of Range/Index above, but adding `required` here. Both need polyfills, why are we saying that we don't expect the consumer to have polyfilled `System.Range`, but do expect them to have polyfilled the required attributes?

Reply:
I do have a strong feeling here. The source package effectively requires consumers to have polyfilled stuff; either it should state what it requires being polyfilled, and should be consistent about the effective TFM that should be polyfilled, or it should not depend on any polyfills being present.